### ISSUE TITLE
Maintaining different versions of force fields

### ISSUE THREAD

**Issue by mattwthompson** [mosdef-hub/foyer#187]
Not sure we've talked about this much, but I think it's worth recording somewhere.

Should version control be handled within `foyer`, in any capacity, or should it be done purely externally, i.e. with git?

Hypothetically (read: something I need done for my own research) I have a force field with recently-updated parameters, but it is useful to have an old version for comparisons. It happens to be buried in the git history, but it's not so easy to grab an old version of a force field just using git. It's not like we have `myff = Forcefield('myff.xml', commit='blah123')`. Grabbing an older version of the file and using just using `myff_old.xml` would work for my purposes, but it doesn't seem to be preferred.

**Comment by ctk3b** [comment]
> Should version control be handled within foyer?

What would you envision `foyer` providing here? At some level you need the copies of the files lying around somewhere.

Off the top of my head you could:
1) have file names that contain the version
2) use git release tags to identify "released" versions

I would probably look to how OpenMM packages their set of supported forcefields and how the OpenFF effort has released versions of their frosst FF

**Comment by ahy3nz** [comment]
If you did something like `myff = Forcefield('myff.xml', commit='blah123')`, doesn't that assume a particular git repo for that force field? What if you tried pulling the xml from the URL of a particular git commit, loading the xml, and then deleting it? That way there's room for the user to specify which repo and which commit to pull the file
```
import requests
url = 'https://raw.githubusercontent.com/choderalab/openmm-forcefields/f94144ea1b9604c06c9831c96311c94422b3179a/charmm/ffxml/charmm36.xml'                                    r = requests.get(url, allow_redirects=True)
open('tempff.xml', 'wb').write(r.content)
```

**Comment by mattwthompson** [comment]
After speaking with @summeraz and @justinGilmer what I have in mind is a way for a user to access old versions of the force field that relies on some sort of metadata encoded in the xml files, maybe like a more useful approach than #153. This definitely could all be handled with git (e.g. maintaining a branch that includes old versions of a force field, or simply using releases), but this would not be accessible to a user who only installs the package from conda and not a development version. It doesn't look like openmm has a default way of handling this, instead just carrying around [different files ](https://github.com/pandegroup/openmm/tree/master/wrappers/python/simtk/openmm/app/data)for older versions. - it seems like their large force fields are already tied together into [single major](
https://github.com/pandegroup/openmm/commit/9d9a3f4fa02ab3b24431c454c5c588c113793fd3#diff-1fc923375846429775eb89387a9688b3) releases, so it might not be such an issue.

What would satisfy the functionality I have in mind is to have some pre-processing in the loading of xml files. Different versions of a force field could be stored into an subdirectory, and the `__init__` function could recursively look into that directory, scanning the files for matching metadata (version, release, git commit, etc.). Of course, this would all be optional, i.e. the `version` in the xml file could be blank, and as long as `Forcefield()` is provided a working path to a file, it could skip over this altogether.

**Comment by mattwthompson** [comment]
Basically yeah, what I have now (and what I think is a sensible way to distribute these files on anaconda) is a python package that has, buried deep in it, a directory containing a main xml file. Old versions would be included in a structure like this (relying on a tag inside the xml file)

```
$ tree
.
├── myff.xml
└── old
    ├── myff_0.xml
    └── myff_1.xml
```

or if for some reason keeping the files named the same is useful (again, tagged in a similar way):
```
$ tree
.
├── myff.xml
└── old
    └── myff_0
        └── myff.xml
    └── myff_1
        └── myff.xml
```

Actually finding the right files might get a little trickier if a "force field" is described by multiple files, i.e. one file for a polymer/protein and another for the water model.

Also, imo, some sort of a version/release attribute to the xml is a useful addition irregardless of these proposed features.

**Comment by mikemhenry** [comment]
I'm not sure what the final implementation would look like, but I'm all for adding more metadata into a forcefield file like a version number.

**Comment by chrisiacovella** [comment]
This is an interesting thread.  I know we had talked about this before, i.e., using version control as a means to keep track of different releases, but never got into the specifics of how that would be handled. @ctk3b you had at one point proposed the idea of just passing a doi to foyer, and having it fetch the forcefield, so I could see that approach working well with passing a commit number (of course there is the issue of multiple forcefield files in a repo, so how do you choose based on a doi...need another parameter).

**Comment by chrisiacovella** [comment]
Oh yeah I forgot zenodo now supports versioning. Each time you add to zenodo you get a unique doi.

**Comment by ctk3b** [comment]
What I had originally proposed and prototyped (there might be a branch floating around) was:

1) one git repo per force field
2) every "published" version of a forcefield gets a git release tag (so when you're ready to publish you just tag it the same way you would do a software package release)
3) zenodo autogenerates DOIs when you update release tags

The only missing piece here I think would be an easy way to see the full history of zenodo DOIs for that repo but I assume that's doable/already provided by zenodo?

**Comment by chrisiacovella** [comment]
looks like it is a new doi per release.  But it also gives you a doi to cite all versions that will always resolve to the latest one.

http://help.zenodo.org/#versioning

**Comment by ctk3b** [comment]
Yup so as long as we can easily pull the full history of releases, then you've got the ability to pick the version of the forcefield you want via DOI.

What I don't think this provides you is an easy way to swap out development versions of the FF via DOI. At that point, you might just be better off using the file directly though

**Comment by chrisiacovella** [comment]
Agreed.   One of the updates we started, but haven't finished, for the forcefield repository, includes providing metadata in an xml file (which would automatically generate the readme file).  We should revisit that in the near future to see if we want to add anything else in there to help us keep track of versions and such (or anything we need to add to the xml file itself).  Similarly, I want to say @mikemhenry was also talking about units in the file (while at aiche).  As we had talked about a few months ago, we should add in the units tags like SMIRNOFF has, (maybe the same ones), even if we don't actually do anything with the units themselves (that is, you still must use one set of expected units).  Might be good to add this to the xml parser as a requirement to help enforce best practices.
